File: examples/predicate_operator.cpp
#include <cstddef>
#include <cstdint>
#include <vector>

#include <tsl.hpp>

struct less_than_op {
    template <class Vec>
    typename Vec::mask_type operator()(
        typename tsl::reg_param<Vec>::type left,
        typename tsl::reg_param<Vec>::type right) const {
        return tsl::less_than<Vec>(left, right);
    }
};

struct negative_op {
    template <class Vec>
    typename Vec::mask_type operator()(typename tsl::reg_param<Vec>::type value) const {
        return tsl::less_than<Vec>(
            value,
            tsl::set1<Vec>(static_cast<typename Vec::base_type>(0)));
    }
};

void fill_inputs(
    std::vector<std::int32_t>& left,
    std::vector<std::int32_t>& right) {
    for (std::size_t i = 0; i < left.size(); ++i) {
        left[i] = static_cast<std::int32_t>(static_cast<int>(i % 41) - 20);
        right[i] = static_cast<std::int32_t>(static_cast<int>((i * 5) % 37) - 18);
    }
}

template <class Mask>
bool mask_lane_is_set(Mask mask, std::size_t lane) {
    return ((static_cast<std::uint64_t>(mask) >> lane) & std::uint64_t{1}) != 0;
}

template <std::size_t ParallelN>
bool run_negative_case() {
    constexpr std::size_t count = 1000;
    std::vector<std::int32_t> input(count);
    std::vector<std::int32_t> scratch(count);
    fill_inputs(input, scratch);

    using mask_type = tsl::algo::fixed_integral_mask_type<ParallelN, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::integral_mask_chunk_count<ParallelN, std::int32_t>(count));

    const auto produced = tsl::algo::predicate_unary<
        ParallelN,
        tsl::algo::alignment::unaligned>(
        negative_op{},
        input.data(),
        masks.data(),
        count);
    if (produced != masks.size()) {
        return false;
    }

    for (std::size_t i = 0; i < count; ++i) {
        const std::size_t chunk = i / ParallelN;
        const std::size_t lane = i % ParallelN;
        const bool expected = input[i] < 0;
        if (mask_lane_is_set(masks[chunk], lane) != expected) {
            return false;
        }
    }
    return true;
}

template <std::size_t ParallelN>
bool run_less_than_case() {
    constexpr std::size_t count = 1000;
    std::vector<std::int32_t> left(count);
    std::vector<std::int32_t> right(count);
    fill_inputs(left, right);

    using mask_type = tsl::algo::fixed_integral_mask_type<ParallelN, std::int32_t>;
    std::vector<mask_type> masks(
        tsl::algo::integral_mask_chunk_count<ParallelN, std::int32_t>(count));

    const auto produced = tsl::algo::predicate_binary<
        ParallelN,
        tsl::algo::alignment::unaligned>(
        less_than_op{},
        left.data(),
        right.data(),
        masks.data(),
        count);
    if (produced != masks.size()) {
        return false;
    }

    for (std::size_t i = 0; i < count; ++i) {
        const std::size_t chunk = i / ParallelN;
        const std::size_t lane = i % ParallelN;
        const bool expected = left[i] < right[i];
        if (mask_lane_is_set(masks[chunk], lane) != expected) {
            return false;
        }
    }
    return true;
}

template <std::size_t ParallelN>
bool run_predicate_cases() {
    return run_negative_case<ParallelN>() && run_less_than_case<ParallelN>();
}

int main() {
    if (!run_predicate_cases<1>()) {
        return 1;
    }
    if (!run_predicate_cases<4>()) {
        return 2;
    }
    if (!run_predicate_cases<16>()) {
        return 3;
    }
    return 0;
}

File: examples/rust/src/bin/predicate_operator.rs
use tsl::profile;
use tsl::tsl_algorithm::{IntegralMask, VectorFor};
use tsl::tsl_core::{SimdVector, StaticSimdVector};

struct LessThan;

impl<V> profile::algo::BinaryPredicateKernel<V> for LessThan
where
    V: StaticSimdVector<BaseType = i32> + profile::detail::primitives::Less_thanImpl,
{
    fn test(&mut self, left: V::RegisterType, right: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(left, right)
    }
}

struct Negative;

impl<V> profile::algo::UnaryPredicateKernel<V> for Negative
where
    V: StaticSimdVector<BaseType = i32>
        + profile::detail::primitives::Less_thanImpl
        + profile::detail::primitives::Set1Impl,
{
    fn test(&mut self, value: V::RegisterType) -> V::MaskType {
        profile::less_than::<V>(value, profile::set1::<V>(0))
    }
}

fn fill_inputs(left: &mut [i32], right: &mut [i32]) {
    for (i, (left_value, right_value)) in left.iter_mut().zip(right.iter_mut()).enumerate() {
        *left_value = (i % 41) as i32 - 20;
        *right_value = ((i * 5) % 37) as i32 - 18;
    }
}

fn mask_lane_is_set(mask: u64, lane: usize) -> bool {
    ((mask >> lane) & 1) != 0
}

fn verify_masks<F>(masks: &[u64], lanes: usize, count: usize, mut expected: F)
where
    F: FnMut(usize) -> bool,
{
    for i in 0..count {
        let chunk = i / lanes;
        let lane = i % lanes;
        assert_eq!(mask_lane_is_set(masks[chunk], lane), expected(i));
    }
}

fn verify_unequal_zero_facade<Policy>(policy: Policy)
where
    Policy: VectorFor<profile::algo::Profile, i32>,
    <Policy as VectorFor<profile::algo::Profile, i32>>::Vec:
        StaticSimdVector<BaseType = i32>
            + profile::detail::primitives::Unequal_zeroImpl
            + profile::detail::primitives::LoadImpl<false>,
    profile::algo::Profile: IntegralMask<<Policy as VectorFor<profile::algo::Profile, i32>>::Vec>,
    <<Policy as VectorFor<profile::algo::Profile, i32>>::Vec as SimdVector>::ImaskType: Into<u64>,
{
    let count =
        <<Policy as VectorFor<profile::algo::Profile, i32>>::Vec as StaticSimdVector>::ELEMENT_COUNT;
    let input: Vec<i32> = (0..count).map(|value| value as i32 - 2).collect();

    let values = unsafe {
        profile::load::<<Policy as VectorFor<profile::algo::Profile, i32>>::Vec, false>(
            input.as_ptr(),
        )
    };
    let mask = profile::algo::unequal_zero::<_, i32>(policy, values);
    let bits: u64 = <profile::algo::Profile as IntegralMask<
        <Policy as VectorFor<profile::algo::Profile, i32>>::Vec,
    >>::to_integral(mask)
    .into();

    for (lane, value) in input.iter().enumerate() {
        assert_eq!(mask_lane_is_set(bits, lane), *value != 0);
    }
}

fn verify_less_than_facade<Policy>(policy: Policy)
where
    Policy: VectorFor<profile::algo::Profile, i32> + Copy,
    <Policy as VectorFor<profile::algo::Profile, i32>>::Vec:
        StaticSimdVector<BaseType = i32>
            + profile::detail::primitives::Less_thanImpl
            + profile::detail::primitives::Mask_trueImpl
            + profile::detail::primitives::Mask_binary_andImpl
            + profile::detail::primitives::Mask_binary_notImpl
            + profile::detail::primitives::LoadImpl<false>,
    profile::algo::Profile: IntegralMask<<Policy as VectorFor<profile::algo::Profile, i32>>::Vec>,
    <<Policy as VectorFor<profile::algo::Profile, i32>>::Vec as SimdVector>::MaskType: Copy,
    <<Policy as VectorFor<profile::algo::Profile, i32>>::Vec as SimdVector>::ImaskType: Into<u64>,
{
    let count =
        <<Policy as VectorFor<profile::algo::Profile, i32>>::Vec as StaticSimdVector>::ELEMENT_COUNT;
    let mut left = vec![0i32; count];
    let mut right = vec![0i32; count];
    fill_inputs(&mut left, &mut right);

    let left_values = unsafe {
        profile::load::<<Policy as VectorFor<profile::algo::Profile, i32>>::Vec, false>(
            left.as_ptr(),
        )
    };
    let right_values = unsafe {
        profile::load::<<Policy as VectorFor<profile::algo::Profile, i32>>::Vec, false>(
            right.as_ptr(),
        )
    };
    let mask = profile::algo::less_than::<_, i32>(policy, left_values, right_values);
    let all = profile::algo::mask_true::<_, i32>(policy);
    let active = profile::algo::mask_binary_and::<_, i32>(policy, all, mask);
    let inactive = profile::algo::mask_binary_not::<_, i32>(policy, active);
    let all_bits: u64 = <profile::algo::Profile as IntegralMask<
        <Policy as VectorFor<profile::algo::Profile, i32>>::Vec,
    >>::to_integral(all)
    .into();
    let active_bits: u64 = <profile::algo::Profile as IntegralMask<
        <Policy as VectorFor<profile::algo::Profile, i32>>::Vec,
    >>::to_integral(active)
    .into();
    let inactive_bits: u64 = <profile::algo::Profile as IntegralMask<
        <Policy as VectorFor<profile::algo::Profile, i32>>::Vec,
    >>::to_integral(inactive)
    .into();

    for lane in 0..count {
        let expected = left[lane] < right[lane];
        assert!(mask_lane_is_set(all_bits, lane));
        assert_eq!(mask_lane_is_set(active_bits, lane), expected);
        assert_eq!(mask_lane_is_set(inactive_bits, lane), !expected);
    }
}

fn main() {
    verify_unequal_zero_facade(tsl::dataparallel::fixed::<1>());
    verify_unequal_zero_facade(tsl::dataparallel::generic::<4>());
    verify_less_than_facade(tsl::dataparallel::fixed::<1>());
    verify_less_than_facade(tsl::dataparallel::generic::<4>());

    let mut left = vec![0i32; 1000];
    let mut right = vec![0i32; 1000];
    fill_inputs(&mut left, &mut right);

    macro_rules! run_policy {
        ($policy:expr, $lanes:expr) => {{
            let policy = $policy;
            let mask_count = profile::algo::integral_mask_chunk_count::<_, i32>(policy, left.len());

            let mut unary_masks = vec![0u64; mask_count];
            let mut negative = Negative;
            let produced =
                profile::algo::predicate_unary(policy, &mut negative, &left, &mut unary_masks);
            assert_eq!(produced, unary_masks.len());
            verify_masks(&unary_masks, $lanes, left.len(), |i| left[i] < 0);

            let mut binary_masks = vec![0u64; mask_count];
            let mut less_than = LessThan;
            let produced = profile::algo::predicate_binary(
                policy,
                &mut less_than,
                &left,
                &right,
                &mut binary_masks,
            );
            assert_eq!(produced, binary_masks.len());
            verify_masks(&binary_masks, $lanes, left.len(), |i| left[i] < right[i]);
        }};
    }

    run_policy!(tsl::dataparallel::fixed::<1>(), 1);
    run_policy!(tsl::dataparallel::generic::<4>(), 4);
    run_policy!(tsl::dataparallel::generic::<16>(), 16);
}
